File: benchmarks-QuantumOptics.jl/expect_state.jl
using QuantumOptics
using BenchmarkTools
include("benchmarkutils.jl")

name = "expect_state"

samples = 5
evals = 100
cutoffs = [5000:5000:150000;]

function setup(N)
    b = FockBasis(N-1)
    op = (destroy(b) + create(b))
    psi = Ket(b, ones(ComplexF64, N)/sqrt(N))
    op, psi
end

function f(op, psi)
    expect(op, psi)
end

println("Benchmarking: ", name)
print("Cutoff: ")
checks = Dict{Int, Float64}()
results = []
for N in cutoffs
    print(N, " ")
    op, psi = setup(N)
    checks[N] = abs(f(op, psi))
    t = @belapsed f($op, $psi) samples=samples evals=evals
    push!(results, Dict("N"=>N, "t"=>t))
end
println()

benchmarkutils.check(name, checks)
benchmarkutils.save(name, results)

File: benchmarks-QuTiP/expect_state.py
import qutip as qt
import numpy as np
from benchmarkutils import benchmark

def setup(N):
    op = (qt.destroy(N) + qt.create(N))
    psi = qt.Qobj(np.ones(N, complex)/np.sqrt(N))
    return (op, psi)

def f(op, psi):
    return qt.expect(op, psi)

if __name__ == '__main__':
    benchmark(name    = 'expect_state', 
              f       = f,
              setup   = setup,
              samples = 5,
              evals   = 100,
              cutoffs = range(500, 10001, 500),
              check_f = f)